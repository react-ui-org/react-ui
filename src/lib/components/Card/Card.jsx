import PropTypes from 'prop-types';
import React from 'react';
import { withProviderContext } from '../../provider';
import { classNames } from '../../utils/classNames';
import { getRootColorClassName } from '../_helpers/getRootColorClassName';
import styles from './Card.scss';

export const Card = ({
  children,
  dense,
  disabled,
  id,
  raised,
  color,
}) => (
  <div
    className={classNames(
      styles.root,
      getRootColorClassName(color, styles),
      dense && styles.rootDense,
      raised && styles.rootRaised,
      disabled && styles.isDisabled,
    )}
    id={id}
  >
    {children}
  </div>
);

Card.defaultProps = {
  color: 'light',
  dense: false,
  disabled: false,
  id: undefined,
  raised: false,
};

Card.propTypes = {
  /**
   * Slot for individual card elements that build up the inner layout:
   * * `CardBody`
   * * `CardFooter`
   * * `ScrollView`
   */
  children: PropTypes.node.isRequired,
  /**
   * [Color variant](/foundation/colors#component-colors) to clarify importance and meaning of the card.
   */
  color: PropTypes.oneOf(
    ['primary', 'secondary', 'success', 'warning', 'danger', 'help', 'info', 'note', 'light', 'dark'],
  ),
  /**
   * Make the card more compact.
   */
  dense: PropTypes.bool,
  /**
   * If `true`, the card will be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * ID of the root HTML element.
   */
  id: PropTypes.string,
  /**
   * Add shadow to pull the card above surface.
   */
  raised: PropTypes.bool,
};

export const CardWithContext = withProviderContext(Card, 'Card');

export default CardWithContext;
