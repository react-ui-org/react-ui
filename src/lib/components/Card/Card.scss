// 1. Retain equal card widths in flex and grid layouts independently on their content.

@use "../../styles/theme/colors";
@use "theme";
@use "tools";

.root {
    --rui-local-padding: #{theme.$padding};

    display: flex;
    flex-direction: column;
    min-width: 0; // 1.
    color: var(--rui-local-color);
    border: theme.$border-width solid var(--rui-local-border-color);
    border-radius: theme.$border-radius;
    background-color: var(--rui-local-background-color);
}

.body {
    flex: 1 0 auto;
    padding: var(--rui-local-padding);
}

.footer {
    padding: var(--rui-local-padding);
}

.rootDense {
    --rui-local-padding: #{theme.$dense-padding};
}

.rootRaised {
    box-shadow: theme.$raised-box-shadow;
}

.rootColorPrimary {
    @include tools.color(primary);
}

.rootColorSecondary {
    @include tools.color(secondary);
}

.rootColorSuccess {
    @include tools.color(success);
}

.rootColorWarning {
    @include tools.color(warning);
}

.rootColorDanger {
    @include tools.color(danger);
}

.rootColorHelp {
    @include tools.color(help);
}

.rootColorInfo {
    @include tools.color(info);
}

.rootColorNote {
    @include tools.color(note);
}

.rootColorLight {
    @include tools.color(light);
}

.rootColorDark {
    @include tools.color(dark);
}

.isDisabled {
    background-color: theme.$disabled-background-color;
    opacity: theme.$disabled-opacity;
}
