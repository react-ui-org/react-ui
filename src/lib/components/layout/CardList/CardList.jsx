import PropTypes from 'prop-types';
import React from 'react';
import styles from './CardList.scss';

const CardList = (props) => {
  const {
    children,
    id,
    ...other
  } = props;

  if (!props.children) {
    return null;
  }

  return (
    <div
      id={id}
      className={styles.root}
      {...other}
    >
      {React.Children.map(children, (child) => {
        if (!React.isValidElement(child)) {
          return null;
        }

        return React.cloneElement(child, {
          inList: true,
        });
      })}
    </div>
  );
};

CardList.defaultProps = {
  children: null,
  id: undefined,
};

CardList.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  id: PropTypes.string,
};

export default CardList;
