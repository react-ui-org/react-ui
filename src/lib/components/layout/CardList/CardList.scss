@import '../../../styles/tools/breakpoints';
@import '../../../styles/tools/offset';
@import './theme';

/* autoprefixer grid: off */
.root {
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: auto;
  grid-gap: $card-list-grid-gap;
  margin-bottom: offset(4);

  @include breakpoint-up(xs) {
    grid-template-columns: repeat(auto-fit, minmax($card-list-card-min-width, $card-list-card-max-width));
    justify-content: start;
  }
}
