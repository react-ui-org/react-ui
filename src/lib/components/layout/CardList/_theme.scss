$card-list-card-min-width: var(--rui-card-list-card-min-width);
$card-list-card-max-width: var(--rui-card-list-card-max-width);
$card-list-grid-gap: var(--rui-card-list-grid-gap);
