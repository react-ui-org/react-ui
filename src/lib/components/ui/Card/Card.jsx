import PropTypes from 'prop-types';
import React from 'react';
import getRootColorClassName from '../../../helpers/getRootColorClassName';
import { withProviderContext } from '../../../provider';
import styles from './Card.scss';

export const Card = ({
  children,
  dense,
  disabled,
  id,
  raised,
  color,
}) => (
  <div
    className={[
      styles.root,
      getRootColorClassName(color, styles),
      dense ? styles.rootDense : '',
      raised ? styles.rootRaised : '',
      disabled ? styles.isDisabled : '',
    ].join(' ')}
    id={id}
  >
    {children}
  </div>
);

Card.defaultProps = {
  color: 'light',
  dense: false,
  disabled: false,
  id: undefined,
  raised: false,
};

Card.propTypes = {
  /**
   * Slot for individual card elements that build up the inner layout:
   * * `CardBody`
   * * `CardFooter`
   * * `ScrollView`
   */
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]).isRequired,
  /**
   * [Color variant](/foundation/colors#component-colors) to clarify importance and meaning of the card.
   */
  color: PropTypes.oneOf(
    ['primary', 'secondary', 'success', 'warning', 'danger', 'help', 'info', 'note', 'light', 'dark'],
  ),
  /**
   * Make the card more compact.
   */
  dense: PropTypes.bool,
  /**
   * If `true`, the card will be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * ID of the root HTML element.
   */
  id: PropTypes.string,
  /**
   * Add shadow to pull the card above surface.
   */
  raised: PropTypes.bool,
};

export const CardWithContext = withProviderContext(Card, 'Card');

export default CardWithContext;
