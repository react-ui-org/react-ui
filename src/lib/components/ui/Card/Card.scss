@import 'mixins';

.root {
  @include card();
}

.flat {
  @include card-type(flat);
}

.bordered {
  @include card-type(bordered);
}

.warning {
  @include card-type(warning);
}

.error {
  @include card-type(error);
}

.isRootDisabled {
  @include card-disabled();
}

.isRootInList {
  @include card-in-list();
}

.isRootRaised {
  @include card-raised();
}
