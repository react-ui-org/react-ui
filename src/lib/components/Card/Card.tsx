import React from 'react';
import { withGlobalProps } from '../../provider';
import { transferProps } from '../_helpers/transferProps';
import { classNames } from '../../utils/classNames';
import { getRootColorClassName } from '../_helpers/getRootColorClassName';
import styles from './Card.scss';
import { CardProps } from './Card.types';

export const Card: React.FunctionComponent<CardProps> = ({
  children,
  dense = false,
  disabled = false,
  raised = false,
  color = 'light',
  ...restProps
}) => (
  <div
    {...transferProps(restProps)}
    className={classNames(
      styles.root,
      getRootColorClassName(color, styles),
      dense && styles.isRootDense,
      raised && styles.isRootRaised,
      disabled && styles.isRootDisabled,
    )}
  >
    {children}
  </div>
);

export const CardWithGlobalProps = withGlobalProps(Card, 'Card');

export default CardWithGlobalProps;
