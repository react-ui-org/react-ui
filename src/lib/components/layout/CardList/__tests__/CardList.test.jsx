import React from 'react';
import { shallow } from 'enzyme';
import { Card } from '../../../ui/Card/Card';
import CardList from '../CardList';

describe('rendering', () => {
  it('renders correctly with no children', () => {
    const tree = shallow((
      <CardList />
    ));

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with a single child', () => {
    const tree = shallow((
      <CardList>
        <Card>content</Card>
      </CardList>
    ));

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with multiple children', () => {
    const tree = shallow((
      <CardList>
        <Card>content 1</Card>
        <Card>content 2</Card>
        <Card>content 3</Card>
      </CardList>
    ));

    expect(tree).toMatchSnapshot();
  });
});
