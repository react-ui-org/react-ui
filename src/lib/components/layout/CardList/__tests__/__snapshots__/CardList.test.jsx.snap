// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly with a single child 1`] = `
<div
  className="root"
>
  <Card
    dense={false}
    disabled={false}
    inList={true}
    key=".0"
    raised={false}
    type="flat"
  >
    content
  </Card>
</div>
`;

exports[`rendering renders correctly with multiple children 1`] = `
<div
  className="root"
>
  <Card
    dense={false}
    disabled={false}
    inList={true}
    key=".0"
    raised={false}
    type="flat"
  >
    content 1
  </Card>
  <Card
    dense={false}
    disabled={false}
    inList={true}
    key=".1"
    raised={false}
    type="flat"
  >
    content 2
  </Card>
  <Card
    dense={false}
    disabled={false}
    inList={true}
    key=".2"
    raised={false}
    type="flat"
  >
    content 3
  </Card>
</div>
`;

exports[`rendering renders correctly with no children 1`] = `""`;
